File: src/main/java/gg/uhc/migration/configuration/MessageSenderReader.java
package gg.uhc.migration.configuration;

import com.google.common.collect.Lists;
import gg.uhc.migration.messages.MessageSender;
import gg.uhc.migration.messages.MultiMessageSender;
import gg.uhc.migration.messages.Senders;
import org.bukkit.configuration.ConfigurationSection;
import org.bukkit.configuration.InvalidConfigurationException;

import java.util.List;

public class MessageSenderReader extends ConfigurationReader<MessageSender> {
    @Override
    protected MessageSender read(ConfigurationSection section, String key) throws InvalidConfigurationException {
        if (!section.isString(key)) throw up(WRONG_TYPE, key, "Message Sender Type", section.get(key));

        String all = section.getString(key);

        String[] types = all.split("\\+");

        List<MessageSender> senders = Lists.newArrayList();
        for(String type : types) {
            try {
                Senders sender = Senders.valueOf(type);

                senders.add(sender.sender);
            } catch (IllegalArgumentException e) {
                throw up(WRONG_TYPE, key, "Message Sender Type", type);
            }
        }

        return new MultiMessageSender(senders);
    }
}

File: src/main/java/gg/uhc/migration/configuration/MessageSenderReader.kt
package gg.uhc.migration.configuration

import gg.uhc.migration.messages.MessageSender
import gg.uhc.migration.messages.MultiMessageSender
import gg.uhc.migration.messages.Senders
import org.bukkit.configuration.ConfigurationSection

fun ConfigurationSection.toMessageSender(key: String) : MessageSender {
    if (!this.isString(key))
        throw wrongType(key, "Message Sender Type", this.get(key))

    val types = this
        .getString(key)
        .split('+')
        .map(String::trim)
        .filterNot(String::isEmpty)
        .map(String::toUpperCase)
        .distinct()
        .map {
            try {
                Senders.valueOf(it)
            } catch (e: IllegalArgumentException) {
                throw wrongType(key, "Message Sender Type", it)
            }
        }
        .map { it.sender }

    return MultiMessageSender(types)
}
